zjit/src/json.rs

```diff
@@ -0,0 +1,529 @@
+//! Single file JSON serializer for iongraph output of ZJIT HIR.
+
+use std::{
+    fmt,
+    io::{self, Write},
+};
+
+pub trait Jsonable {
+    fn to_json(&self) -> Json;
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Json {
+    Null,
+    Bool(bool),
+    Integer(i64),
+    UnsignedInteger(usize),
+    Floating(f64),
+    String(String),
+    Array(Vec<Json>),
+    Object(Vec<(String, Json)>),
+}
+
+impl Json {
+    /// Convenience method for constructing a JSON array.
+    pub fn array<I, T>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        T: Into<Json>,
+    {
+        Json::Array(iter.into_iter().map(Into::into).collect())
+    }
+
+    pub fn empty_array() -> Self {
+        Json::Array(Vec::new())
+    }
+
+    pub fn object() -> JsonObjectBuilder {
+        JsonObjectBuilder::new()
+    }
+
+    pub fn marshall<W: Write>(&self, writer: &mut W) -> JsonResult<()> {
+        match self {
+            Json::Null => writer.write_all(b"null"),
+            Json::Bool(b) => writer.write_all(if *b { b"true" } else { b"false" }),
+            Json::Integer(i) => write!(writer, "{}", i),
+            Json::UnsignedInteger(u) => write!(writer, "{}", u),
+            Json::Floating(f) => write!(writer, "{}", f),
+            Json::String(s) => return Self::write_str(writer, s),
+            Json::Array(jsons) => return Self::write_array(writer, jsons),
+            Json::Object(map) => return Self::write_object(writer, map),
+        }?;
+        Ok(())
+    }
+
+    pub fn write_str<W: Write>(writer: &mut W, s: &str) -> JsonResult<()> {
+        writer.write_all(b"\"")?;
+
+        for ch in s.chars() {
+            match ch {
+                '"' => write!(writer, "\\\"")?,
+                '\\' => write!(writer, "\\\\")?,
+                '/' => write!(writer, "\\/")?,
+                // The following characters are control, but have a canonical representation.
+                '\n' => write!(writer, "\\n")?,
+                '\r' => write!(writer, "\\r")?,
+                '\t' => write!(writer, "\\t")?,
+                '\x08' => write!(writer, "\\b")?,
+                '\x0C' => write!(writer, "\\f")?,
+                ch if ch.is_control() => {
+                    let code_point = ch as u32;
+                    write!(writer, "\\u{:04X}", code_point)?
+                }
+                _ => write!(writer, "{}", ch)?,
+            };
+        }
+
+        writer.write_all(b"\"")?;
+        Ok(())
+    }
+
+    pub fn write_array<W: Write>(writer: &mut W, jsons: &[Json]) -> JsonResult<()> {
+        writer.write_all(b"[")?;
+        for (i, item) in jsons.iter().enumerate() {
```

Comment:
https://www.cs.tufts.edu/comp/40-2011f/idioms.html#toc8 (as we do elsewhere)


`let mut sep = ""`, ...